validation/superchain-version_test.go

```diff
@@ -80,69 +57,107 @@ func TestSuperchainWideContractVersions(t *testing.T) {
-			require.NoErrorf(t, err, "%s/%s.%s.version= UNSPECIFIED", chain.Superchain, chain.Name, localizedContractName)
-			checkSemverForContract(t, localizedContractName, &contractAddress, client, desiredSemver)
-		}
+		isFPAC := chain.ChainID == 10 || chain.ChainID == 11155420 || chain.ChainID == 11155421 // TODO don't hardcode this
+		versions, err := getContractVersionsFromChain(*Addresses[chain.ChainID], client, isFPAC)
+		require.NoError(t, err)
+		matches, err := findOPContractTag(versions)
+		require.NoError(t, err)
+		require.NotEmpty(t, matches)
 	}
 
 	for chainID, chain := range OPChains {
+		chainID, chain := chainID, chain
 		t.Run(perChainTestName(chain), func(t *testing.T) {
+			t.Parallel()
 			if isExcluded[chainID] {
 				t.Skipf("chain %d: EXCLUDED from contract version validation", chainID)
 			}
 			RunOnStandardAndStandardCandidateChains(t, *chain)
-			checkOPChainSatisfiesSemver(t, chain)
+			checkOPChainMatchesATag(t, chain)
 		})
 	}
 }
 
+// getContractVersionsFromChain pulls the appropriate contract versions (depending on the isFPAC argument) from chain
+// using the supplied client (calling the version() method for each contract). It does this concurrently.
+func getContractVersionsFromChain(list AddressList, client *ethclient.Client, isFPAC bool) (ContractVersions, error) {
+	// build up list of contracts to check
+	contractsToCheck := []string{
+		"L1CrossDomainMessengerProxy",
+		"L1ERC721BridgeProxy",
+		"L1StandardBridgeProxy",
+		"OptimismMintableERC20FactoryProxy",
+		"OptimismPortalProxy",
+		"SystemConfigProxy",
+	}
+
+	if !isFPAC {
+		contractsToCheck = append(contractsToCheck, "L2OutputOracleProxy")
+	} else {
+		contractsToCheck = append(contractsToCheck,
+			"AnchorStateRegistryProxy",
+			"DelayedWETHProxy",
+			"DisputeGameFactoryProxy",
+			"FaultDisputeGame",
+			"MIPS",
+			"PermissionedDisputeGame",
+			"PreimageOracle",
+		)
+	}
+
+	// Prepare a concurrency-safe object to store version information in, and
+	// spin up a goroutine for each contract we are checking (to speed things up).
+	results := new(sync.Map)
+
+	getVersionAsync := func(contractAddress Address, results *sync.Map, key string, wg *sync.WaitGroup) {
+		r, err := getVersion(context.Background(), common.Address(contractAddress), client)
+		if err != nil {
+			panic(err)
+		}
+		results.Store(key, r)
+		wg.Done()
+	}
+
+	wg := new(sync.WaitGroup)
+
+	for _, contractAddress := range contractsToCheck {
+		a, err := list.AddressFor(contractAddress)
+		if err != nil {
+			return ContractVersions{}, err
+		}
+		wg.Add(1)
+		go getVersionAsync(a, results, contractAddress, wg)
+	}
+
+	wg.Wait()
+
+	// use reflection to convert results mapping into a ContractVersions object
+	// without resorting to boilerplate code.
+	cv := ContractVersions{}
+	results.Range(func(k, v any) bool {
+		s := reflect.ValueOf(cv)
+		for i := 0; i < s.NumField(); i++ {
+			if s.Type().Field(i).Name == k || s.Type().Field(i).Name+"Proxy" == k {
```

Comment:
Can you explain what the `|| s.Type().Field(i).Name+"Proxy" == k` portion of this code is doing?

Reply:
[b3f4558](https://github.com/ethereum-optimism/superchain-registry/pull/297/commits/b3f4558a31ac16ee883642b3cc2d1a93e3decc79)